### ISSUE TITLE
I am trying to add DeepSeekV2Moe model

### ISSUE THREAD

**Issue by qiyuxinlin** [turboderp/exllamav2#554]
Hi,turboderp! I am trying to add DeepSeekV2Moe model in exllamav2, I change torch code but the output looks like only the last token of the input can be seen, It feels like there is an error in the cache location. But I check the past_len, I have not modified the cache-related code, and there is no problem with the cache's current_len. I tried to compare it with the intermediate results from the transformers. However, I found that since transformers use bfloat16 and exllamav2 uses float16, there is a difference in accuracy after the model passes through the first layer of input_layernorm. It is completely different when selecting expert later. 
Have you ever encountered such a situation before where the questions and answers do not correspond?

**Comment by qiyuxinlin** [comment]
![image](https://github.com/user-attachments/assets/685e6110-fc55-41ed-a521-5e34655553b4)
my input is "There's no such thing as", it seems like the model only see the "as"

**Comment by qiyuxinlin** [comment]
Unexpectedly, it is really caused by the data type. For linear and RMSNorm, during the calculation process, I used bfloat16 to calculate the intermediate variables, and the results were converted to float16. Now it works.
![image](https://github.com/user-attachments/assets/426f22c3-8842-4bcd-8260-e857669551d9)

**Comment by turboderp** [comment]
This looks promising. Where did you implement the shared expert?

**Comment by qiyuxinlin** [comment]
Consistent with transformers, I added share_experts at the end of the forward function inside your ExLlamaV2MoEMLP class
```
    def forward_torch(self,
                      hidden_states: torch.Tensor,
                      cache = None,
                      attn_params = None,
                      past_len = None,
                      intermediates = False,
                      loras: list[ExLlamaV2Lora] | None = None,
                      **kwargs) -> torch.Tensor | dict[str: torch.Tensor]:

        residual = hidden_states

        batch_size, sequence_length, hidden_dim = hidden_states.shape
        hidden_states = hidden_states.view(-1, hidden_dim)

        # Layernorm

        hidden_states = self.post_attention_layernorm.forward(hidden_states)
        if intermediates: result = { "post_norm": hidden_states }

        # Get router logits

        router_logits = self.gate.forward(hidden_states, loras = loras)  #[:, :self.num_experts]

        # Get routing weights and select top K experts

        routing_weights = F.softmax(router_logits, dim = -1, dtype = torch.float)
        routing_weights, selected_experts = torch.topk(routing_weights, self.num_experts_per_token, dim = -1)
        if self.norm_topk_prob:
            routing_weights /= routing_weights.sum(dim = -1, keepdim = True)
        routing_weights = routing_weights.to(hidden_states.dtype)

        final_hidden_states = torch.zeros((batch_size * sequence_length, hidden_dim), dtype = hidden_states.dtype, device = hidden_states.device)

        expert_mask = torch.nn.functional.one_hot(selected_experts, num_classes = self.num_experts).permute(2, 1, 0)

        for expert_idx in range(self.num_experts):
            idx, top_x = torch.where(expert_mask[expert_idx])

            if top_x.shape[0] == 0: continue  # Skip experts that weren't selected at all

            top_x_list = top_x.tolist()
            idx_list = idx.tolist()

            current_state = hidden_states[None, top_x_list].reshape(-1, hidden_dim)

            gate = self.w1[expert_idx].forward(current_state, loras = loras)
            up = self.w3[expert_idx].forward(current_state, loras = loras)

            if self.model.config.arch.mlp_act_func == "silu":
                current_hidden_states = F.silu(gate)
            elif self.model.config.arch.mlp_act_func == "gelu":
                current_hidden_states = F.gelu(gate)
            current_hidden_states *= up
            if intermediates: result[f"pre_down.{expert_idx}"] = current_hidden_states

            current_hidden_states = self.w2[expert_idx].forward(current_hidden_states, loras = loras)
            current_hidden_states *= routing_weights[top_x_list, idx_list, None]

            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
        
        ### Qwen2 share_experts
        if self.model.config.arch.mlp_act_func == "silu":
            shared_expert_output = self.shared_expert_w2.forward(F.silu(self.shared_expert_w1.forward(hidden_states)) * self.shared_expert_w3.forward(hidden_states))
        else:
            shared_expert_output = self.shared_expert_w2.forward(F.gelu(self.shared_expert_w1.forward(hidden_states)) * self.shared_expert_w3.forward(hidden_states))
        shared_expert_output = F.sigmoid(self.shared_expert_gate.forward(hidden_states)) * shared_expert_output
        final_hidden_states = final_hidden_states + shared_expert_output
        ###
        final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
        final_hidden_states += residual

        if intermediates:
            result["hidden_states"] = final_hidden_states
            return result
        else:
            return final_hidden_states
```

I don’t understand why the accuracy difference of RMSNorm causes the model output to be so abnormal. When loading weight, your code processes it as float16, while transformers processes it as bfloat16. Have you encountered similar problems before

**Comment by turboderp** [comment]
It's possible there's an overflow happening. The custom RMSNorm kernel shouldn't have this issue since it applies the weights in FP32 before downcasting which gives the same range as BF16 with more precision. So this should also work:

```python
        output_shape = hidden_states.shape
        hidden_states = hidden_states.view(-1, hidden_states.shape[-1])
        norm = torch.empty_like(hidden_states, dtype = torch.half)
        ext_c.rms_norm(hidden_states, self.weight, norm, self.variance_epsilon)
        hidden_states = norm.view(output_shape)
```

**Comment by qiyuxinlin** [comment]
I will learn how it is implemented in your RMSnorm kernel over the weekend. Your forward function produced the following answer. This is why I changed it to bfloat16 for calculation.
![image](https://github.com/user-attachments/assets/3ee5ed51-bfbc-44e8-85cf-da27553ff1ff)

**Comment by linchen111** [comment]
amazing work

**Comment by qiyuxinlin** [comment]
I'm sorry that my previous conclusion was wrong. I tried to understand your RMSNorm code. In your 0.1.7 RMSNorm kernel, running the branch of half summation calculation will lead to wrong results, while running the branch of float summation calculation is no problem, but these two branches actually do the same thing. Finally, I found that because I did torch.split on hidden_states, the memory addresses of hidden_states were discontinuous, but 
``hidden_states.view(-1, hidden_states.shape[-1]) `` could still be executed. , the result I got using is_contiguous() is also False, which is amazing. 
I checked the official documentation of torch. torch.view cannot execute on variables with discontinuous memory addresses.
